core/metacling/src/TCling.cxx

```diff
@@ -6149,116 +6155,134 @@ static bool LookupNormalSymbols(llvm::object::ObjectFile *RealSoFile, const std:
+      // libraries. However, it seems to be redundant for the moment as we do
+      // not strictly require symbols from those sections. Enable after checking
+      // performance!
+      if (LookupNormalSymbols(LibName, mangled_name)) {
+         sQueriedLibraries.push_back(P);
+         return LibName;
+      }
+   }
 
-      auto RealSoFile = SoFile.get().getBinary();
+   if (!searchSystem)
+      return "";
 
-      if (LookupNormalSymbols(RealSoFile, name_in_so, LibName)) {
-         if (gSystem->Load(LibName.c_str(), "", false) < 0)
-            Error("LazyFunctionCreatorAutoloadForModule", "Failed to load library %s", LibName.c_str());
+   // Lookup in non-system libraries failed. Expand the search to the system.
+   if (sFirstSystemLibrary) {
+      TCling__FindLoadedLibraries(sSysLibraries, sPaths, *interp, /* searchSystem */ true);
+      sFirstSystemLibrary = false;
+   }
 
-         sSysLibraries.erase(std::remove(sSysLibraries.begin(), sSysLibraries.end(), P), sSysLibraries.end());
-         void* addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(mangled_name.c_str());
-         return addr;
-      }
+   for (const LibraryPath &P : sSysLibraries) {
+      const std::string LibName = GetLibFileName(P, sPaths);
 
-      // Lookup for dynamic symbols
-      auto sections = RealSoFile->sections();
-      for (auto section : sections) {
-         llvm::StringRef sectionName;
-         section.getName(sectionName);
-
-         // .dynstr contains string of dynamic symbols
-         if (sectionName == ".dynstr") {
-            llvm::StringRef dContents;
-            section.getContents(dContents);
-            // If this library contains mangled name
-            if (dContents.contains(mangled_name)) {
-               if (gSystem->Load(LibName.c_str(), "", false) < 0)
-                  Error("LazyFunctionCreatorAutoloadForModule", "Failed to load library %s", LibName.c_str());
-
-               // Delete a loaded library from sLibraries cache.
-               sSysLibraries.erase(std::remove(sSysLibraries.begin(), sSysLibraries.end(), P), sSysLibraries.end());
-               void* addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(mangled_name.c_str());
-               return addr;
-            }
-         }
+      if (LookupNormalSymbols(LibName, mangled_name) || LookupDynamicSymbols(LibName, mangled_name)) {
+         sQueriedLibraries.push_back(P);
+         return LibName;
       }
    }
 
-   // Lookup failed!!!!
-   return nullptr;
+   return ""; // Search found no match.
+}
+
+static void* LazyFunctionCreatorAutoloadForModule(const std::string& mangled_name,
+                                                  cling::Interpreter *interp) {
+   std::string LibName = SearchSharedLibraries(mangled_name, interp);
+   if (LibName.empty())
+      return nullptr;
+
+   if (gSystem->Load(LibName.c_str(), "", false) < 0)
+      Error("LazyFunctionCreatorAutoloadForModule", "Failed to load library %s", LibName.c_str());
```

Comment:
Can you rename the function of the message to contain `TCling__` so people know where this is coming from? (Same comment applies to other static functions with `Error` calls.